gadget/install/install.go

```diff
@@ -377,6 +378,160 @@ func Run(model gadget.Model, gadgetRoot, kernelRoot, bootDevice string, options
+// device name like sda1 or nvme0n1p2.
+func partIndexFromPartName(part string) (uint64, error) {
+	partRunes := []rune(part)
+	i := len(partRunes)
+	for ; i > 0; i-- {
+		if !unicode.IsDigit(partRunes[i-1]) {
+			break
+		}
+	}
+
+	partIdxStr := string(partRunes[i:])
+	if partIdxStr == "" {
+		return 0, fmt.Errorf("partition index not found in %q", part)
+	}
+
+	return strconv.ParseUint(partIdxStr, 10, 64)
+}
+
+// structureFromIndex returns the OnDiskStructure for an index from
+// the disk volume.
+func structureFromIndex(diskVol *gadget.OnDiskVolume, idx uint64) (*gadget.OnDiskStructure, error) {
+	for _, p := range diskVol.Structure {
+		if uint64(p.DiskIndex) == idx {
+			return &p, nil
+		}
+	}
+
+	return nil, fmt.Errorf("cannot find partition index %d", idx)
+}
+
+// laidOutStructForDiskStruct searches for the laid out structure that
+// matches a given OnDiskStructure.
+func laidOutStructForDiskStruct(laidVols map[string]*gadget.LaidOutVolume, gadgetVolName string, onDiskStruct *gadget.OnDiskStructure) (
+	*gadget.LaidOutStructure, error) {
+
+	for _, laidVol := range laidVols {
+		// Check that this is the right volume
+		if laidVol.Name != gadgetVolName {
+			continue
+		}
+		for _, laidStruct := range laidVol.LaidOutStructure {
+			if onDiskStruct.Name == laidStruct.Name {
+				return &laidStruct, nil
+			}
+		}
+	}
+
+	return nil, fmt.Errorf("cannot find laid out structure for %q", onDiskStruct.Name)
+}
+
+// sysfsPathForBlockDevice returns the sysfs path for a block device.
+func sysfsPathForBlockDevice(device string) (string, error) {
+	partPath, err := os.Readlink(filepath.Join("/sys/class/block", device))
+	if err != nil {
+		return "", err
+	}
+	// Remove initial ../../ and make path absolute
+	return filepath.Join("/sys", partPath[6:]), nil
+}
+
+// onDiskVolumeFromPartitionSysfsPath creates an OnDiskVolume that
+// matches the disk that contains the given partition sysfs path
+func onDiskVolumeFromPartitionSysfsPath(partPath string) (*gadget.OnDiskVolume, error) {
+	// Removing the last component will give us the disk path
+	diskPath := filepath.Dir(partPath)
+	disk, err := disks.DiskFromDevicePath(diskPath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot retrieve disk information for %q: %v", partPath, err)
+	}
+	onDiskVol, err := gadget.OnDiskVolumeFromDisk(disk)
+	if err != nil {
+		return nil, fmt.Errorf("cannot retrieve on disk volume for %q: %v", partPath, err)
+	}
+
+	return onDiskVol, nil
+}
+
+// buildOnDiskStruct obtains the current on disk structure from a
+// sysfs partition path string. Also, we take the laid out information
+// from laidOutVols and insert it in the disk structure.
+func buildOnDiskStruct(onDiskVol *gadget.OnDiskVolume, partPath string, laidOutVols map[string]*gadget.LaidOutVolume, gadgetVolName string) (*gadget.OnDiskStructure, error) {
+	partName := filepath.Base(partPath)
+	partIdx, err := partIndexFromPartName(partName)
+	if err != nil {
```

Comment:
can't we read  partPath/partition instead ? or pass around the Disk we obtain onDiskVolumeFromPartitionSysfsPath and match through its Partitions?